src/utils/batch-operate/SessionQueue.ts

```diff
@@ -0,0 +1,200 @@
+	private latestProcessingPromise: Promise<void> | null = null;
+	private processingPromisesCount = 0;
+
+	private client: AirtopClient;
+	private sessionPool: string[] = [];
+
+	constructor({
+		maxConcurrentSessions,
+		runEmitter,
+		maxWindowsPerSession,
+		initialBatches,
+		operation,
+		client,
+		sessionConfig,
+		onError,
+	}: {
+		maxConcurrentSessions: number;
+		runEmitter: EventEmitter;
+		maxWindowsPerSession: number;
+		initialBatches: BatchOperationUrl[][];
+		operation: (input: BatchOperationInput) => Promise<BatchOperationResponse | undefined>;
+		client: AirtopClient;
+		sessionConfig?: AirtopSessionConfigV1;
+		onError?: (error: BatchOperationError) => Promise<void>;
+	}) {
+		if (
+			!Number.isInteger(maxConcurrentSessions) ||
+			maxConcurrentSessions <= 0
+		) {
+			throw new Error("maxConcurrentSessions must be a positive integer");
+		}
+
+		this.maxConcurrentSessions = maxConcurrentSessions;
+		this.runEmitter = runEmitter;
+		this.maxWindowsPerSession = maxWindowsPerSession;
+		this.sessionConfig = sessionConfig;
+		this.initialBatches = initialBatches;
+		this.operation = operation;
+		this.onError = onError;
+		this.latestProcessingPromise = null;
+
+		this.client = client;
+	}
+
+	public async addUrlsToBatchQueue(
+		newBatch: BatchOperationUrl[],
+	): Promise<void> {
+		// Distribute new URLs across batches
+		const newBatches = distributeUrlsToBatches(
+			newBatch,
+			this.maxConcurrentSessions,
+		);
+
+		this.client.log(
+			`Adding new batches to queue: ${JSON.stringify(newBatches)}`,
+		);
+		this.batchQueue.push(...newBatches);
+
+		// Update existing processing promise
+		this.processingPromisesCount++;
+		this.latestProcessingPromise = this.processPendingBatches();
+	}
+
+	public async processInitialBatches(): Promise<void> {
+		this.batchQueue = [...this.initialBatches];
+		this.processingPromisesCount++;
+		this.latestProcessingPromise = this.processPendingBatches();
+		await this.latestProcessingPromise;
+	}
+
+	public async waitForProcessingToComplete(): Promise<void> {
+		while (this.processingPromisesCount > 0) {
+			await this.latestProcessingPromise;
+		}
+
+		await this.terminateAllSessions();
+	}
+
+	private async terminateAllSessions(): Promise<void> {
+		for (const sessionId of this.sessionPool) {
+			await this.client.sessions.terminate(sessionId);
+		}
+	}
+
+	private async processPendingBatches(): Promise<void> {
+		while (this.batchQueue.length > 0) {
+			// Wait for any session to complete before starting a new one
+			if (this.activePromises.length >= this.maxConcurrentSessions) {
+				await Promise.race(this.activePromises);
+				continue;
+			}
+
+			const batch = this.batchQueue.shift();
+			if (!batch || batch.length === 0) break;
+
+			const promise = (async () => {
+				let sessionId: string | undefined;
+				try {
+					// Check if there's an available session in the pool
+					if (this.sessionPool.length > 0) {
+						sessionId = this.sessionPool.pop();
```

Comment:
This seems slightly risky to me without a mutex, where basically you could potentially end up with two promises doing the sessionPool.length check and then popping. If there was only 1 session in the pool, one of the promises would get an undefined sessionId and would then throw an error below. Maybe I'm being overly paranoid.

Reply:
A mutex does seem like a good idea.  One (potential) complication is that I'm not sure if/how we can manage custom dependencies in this repo.  To do this we'd need to add a package like `async-mutex` to our `package.json`, and without adding that entire file to `.fernignore` I'm not sure how we could keep it from getting overwritten.  Fern might have a way to support this and I'm happy to ask if you all feel like it's worth it.

On the flip side, we should be catching errors already, so if a user did hit this issue they would have all the tools available to retry the failed operation.